### PR TITLE
Issue #2742 Change to htpasswd addon for v3.10

### LINKED ISSUE #2742: Apply of default addon htpasswd-identity-provider on running minishift instance throws error
### General information

  * Minishift version: v1.23.0
  * OS: Linux / macOS / Windows
  * Hypervisor: KVM / xhyve / Hyper-V / VirtualBox

### Steps to reproduce

  1. minishift start
```
$ ./minishift start
```
  2. minishift addon apply htpasswd-identity-provider --addon-env USER_PASSWORD=test123
```
$ ./minishift addon apply htpasswd-identity-provider --addon-env USER_PASSWORD=test123
-- Applying addon 'htpasswd-identity-provider':..Error applying the add-on: Error executing command 'ssh command error:
command : sudo cp -fp /var/lib/minishift/openshift.local.config/master/master-config.yaml /var/lib/minishift/openshift.local.config/master/master-config-htpasswd.yaml
err     : exit status 1
output  : cp: cannot stat '/var/lib/minishift/openshift.local.config/master/master-config.yaml': No such file or directory
':
```
  3. 
  4. 

### Expected
Should be applied 

### Actual
Throws error

### Logs

You can start Minishift with `minishift start --show-libmachine-logs -v5` to collect logs.
Please consider posting this on http://gist.github.com/ and post the link in the issue.

### PR DESCRIPTION
**Pull request by praveenkumar**
Fixes  #2742 

Steps to test the pull request

  1. Use this PR and apply htpasswd-identity-provider
  2. Go to web console and try to login with different user => this should fail.
  3. login with same user with the password you provide during the apply of the addon (default: developer) => should successful.

### CODE REVIEW

**Comment by praveenkumar** [comment]
retest this please.

**Comment by amitkrout** [comment]
@praveenkumar 
```
$ ./minishift addons list
Error installing default add-ons : Asset htpasswd-identity-provider not found
```

**Comment by praveenkumar** [comment]
@amitkrout how did you get it? I compiled this PR and following steps work for me.

```
$ ./minishift addon install --defaults
Default add-ons 'anyuid, admin-user, xpaas, registry-route, che, htpasswd-identity-provider' installed
[prkumar@local bin]$ ./minishift addon list
- admin-user			 : disabled	P(0)
- anyuid			 : disabled	P(0)
- che				 : disabled	P(0)
- htpasswd-identity-provider	 : disabled	P(0)
- registry-route		 : disabled	P(0)
- xpaas				 : disabled	P(0)

```

**Comment by amitkrout** [comment]
@praveenkumar Take the compiled binary of this pr and run ```./minishift addons list``` on fresh setup.

**Comment by praveenkumar** [comment]
>  Take the compiled binary of this pr and run ./minishift addons list on fresh setup, i mean no old config (.minishift) should be in the setup

@amitkrout doesn't hit this issue.

```
$ curl -sL -O http://artifacts.ci.centos.org/minishift/minishift/pr/2764/linux-amd64/minishift
$ chmod +x minishift 

$ ./minishift addon list
- admin-user			 : disabled	P(0)
- anyuid			 : disabled	P(0)
- che				 : disabled	P(0)
- htpasswd-identity-provider	 : disabled	P(0)
- registry-route		 : disabled	P(0)
- xpaas				 : disabled	P(0)

$ ./minishift version
minishift v1.23.0+c266bc2
```

File: addons/htpasswd-identity-provider/htpasswd-identity-provider.addon [hunk 1]
```diff
@@ -1,30 +1,44 @@
 # Name: htpasswd-identity-provider
 # Description: Configures minishift to use HTPasswdIdentityProvider
 # Url: https://docs.okd.io/3.9/install_config/configuring_authentication.html#HTPasswdPasswordIdentityProvider
-# Var-Defaults: USERNAME=developer,USER_PASSWORD=developer,MINISHIFT_DATA_HOME=/var/lib/minishift,CONFIG_LOCATION=openshift.local.config/master,MASTER_CONFIG_FILE=master-config.yaml,ORIGIN_DATA_HOME=/var/lib/origin
-# Required-Vars: USER_PASSWORD,USERNAME,CONFIG_LOCATION
+# Var-Defaults: USERNAME=developer,USER_PASSWORD=developer,MINISHIFT_DATA_HOME=/var/lib/minishift,CONFIG_LOCATION_OPENSHIFT_API=base/openshift-apiserver,CONFIG_LOCATION_KUBERNETES_API=base/kube-apiserver,MASTER_CONFIG_FILE=master-config.yaml
+# Required-Vars: USER_PASSWORD,USERNAME,CONFIG_LOCATION_OPENSHIFT_API
```
Comment (LalatenduMohanty) [inline addons/htpasswd-identity-provider/htpasswd-identity-provider.addon]: @praveenkumar Can we add version constraint here
```
# OpenShift-Version: >=3.10.10                                                            

```

Comment (LalatenduMohanty) [reply addons/htpasswd-identity-provider/htpasswd-identity-provider.addon]: As it will not work for 3.9 or below versions.

Comment (praveenkumar) [reply addons/htpasswd-identity-provider/htpasswd-identity-provider.addon]: Yes, that we should do :+1:

File: addons/htpasswd-identity-provider/htpasswd-identity-provider.addon [hunk 2]
```diff
@@ -1,30 +1,45 @@
...
 
 # create users.htpasswd file 
-ssh sudo touch #{MINISHIFT_DATA_HOME}/#{CONFIG_LOCATION}/users.htpasswd
+ssh sudo touch #{MINISHIFT_DATA_HOME}/#{CONFIG_LOCATION_OPENSHIFT_API}/users.htpasswd
+ssh sudo touch #{MINISHIFT_DATA_HOME}/#{CONFIG_LOCATION_KUBERNETES_API}/users.htpasswd
 
-# add the default developer user to file
-docker exec -t origin /usr/bin/bash -c "htpasswd -b #{ORIGIN_DATA_HOME}/#{CONFIG_LOCATION}/users.htpasswd #{USERNAME} #{USER_PASSWORD}"
+# copy htpasswd from container to VM
+ssh sudo docker cp origin:/usr/bin/htpasswd /usr/local/bin/
```
Comment (gbraad) [inline addons/htpasswd-identity-provider/htpasswd-identity-provider.addon]: this is risky...

we should be installing to the tmp folder on `/var/sda1/tmp` instead and have this folder (or `/var/sda1/bin` part of `$PATH` to overcome this.

File: addons/htpasswd-identity-provider/htpasswd-identity-provider.addon [hunk 3]
```diff
@@ -1,30 +1,45 @@
...
 
-# Patch the master configuration to use HTPasswdPasswordIdentityProvider
-ssh sudo grep "HTPasswdPasswordIdentityProvider" #{MINISHIFT_DATA_HOME}/#{CONFIG_LOCATION}/#{MASTER_CONFIG_FILE} > /dev/null || sudo #{MINISHIFT_DATA_HOME}/bin/oc ex config patch #{MINISHIFT_DATA_HOME}/#{CONFIG_LOCATION}/master-config-htpasswd.yaml --patch='{"oauthConfig": {"identityProviders": [ {"challenge": true,"login": true,"mappingMethod": "add","name": "htpasswd","provider": {"apiVersion": "v1","kind": "HTPasswdPasswordIdentityProvider","file": "users.htpasswd"}}]}}' > #{MINISHIFT_DATA_HOME}/#{CONFIG_LOCATION}/#{MASTER_CONFIG_FILE}
+# Add user to users.htpasswd file
+ssh sudo htpasswd -b #{MINISHIFT_DATA_HOME}/#{CONFIG_LOCATION_OPENSHIFT_API}/users.htpasswd #{USERNAME} #{USER_PASSWORD}
```
Comment (gbraad) [inline addons/htpasswd-identity-provider/htpasswd-identity-provider.addon]: can we invoke `htpasswd` using an absolute path for now (and place it somewhere else)?

File: addons/htpasswd-identity-provider/htpasswd-identity-provider.addon [hunk 4]
```diff
@@ -1,30 +1,45 @@
...
+
+ssh until curl -f -k https://#{ip}:8443/healthz;do sleep 1;done
 
 # remove the prepatch file
-ssh sudo rm -f #{MINISHIFT_DATA_HOME}/#{CONFIG_LOCATION}/master-config-htpasswd.yaml
+ssh sudo rm -f #{MINISHIFT_DATA_HOME}/#{CONFIG_LOCATION_OPENSHIFT_API}/master-config-htpasswd.yaml
+ssh sudo rm -f #{MINISHIFT_DATA_HOME}/#{CONFIG_LOCATION_KUBERNETES_API}/master-config-htpasswd.yaml
 
-echo 'Successfully installed addon htpasswd identity provider'
+echo -- Successfully installed addon htpasswd identity provider ... OK
```
Comment (gbraad) [inline addons/htpasswd-identity-provider/htpasswd-identity-provider.addon]: I appreciate the fakery here ;-)

File: addons/htpasswd-identity-provider/htpasswd-identity-provider.addon.remove [hunk 5]
```diff
@@ -1,20 +1,32 @@
...
+docker stop $(docker ps -l -q --filter "label=io.kubernetes.container.name=api")
+
+ssh until curl -f -k https://#{ip}:8443/healthz;do sleep 1;done
+
+# remove the prepatch file
+ssh sudo rm -f #{MINISHIFT_DATA_HOME}/#{CONFIG_LOCATION_OPENSHIFT_API}/master-config-htpasswd.yaml
+ssh sudo rm -f #{MINISHIFT_DATA_HOME}/#{CONFIG_LOCATION_KUBERNETES_API}/master-config-htpasswd.yaml
 
-echo 'Successfully removed addon htpasswd-identity-provider'
+echo -- Successfully removed addon htpasswd-identity-provider ... OK
```
Comment (gbraad) [inline addons/htpasswd-identity-provider/htpasswd-identity-provider.addon.remove]: `htpasswd` is not removed from the filesystem...

Comment (praveenkumar) [reply addons/htpasswd-identity-provider/htpasswd-identity-provider.addon.remove]: @gbraad All done now.

**Pull request merged**